content/posts/2018-10-21--complex-command-handler-in-javascript.md

```diff
@@ -0,0 +1,352 @@
+
+We can clearly see that this implementation is better and easier to analyze, so let's use it instead of the previous one.
+
+Since it went so well for `handleInSequence`, let's now get back to implementing `handleInParallel` but this time using `async`/`await` approach. In this case it doesn't result in an easy to read code, though.
+
+First we can wonder why we can't simply call `handle` method for each command inside a `for ... of` loop and `await` for them to be resolved? The problem is that this will not yield the expected result --- note that we've already used this approach when we implemented `handleInSequence` method! The reason is that `await` would pause the execution of our method for every single `handle` call. This would prevent handling of subsequent commands for even getting started and instead of being executed in parallel they would be run in sequence.
+
+Therefore we need to initiate handling of all commands first, and only then `await` for their results.    
+
+```js
+async handleInParallel(commands) {
+    const promises = commands.map((command) => {
+        return this.handle(command);
+    });
+
+
+    for (const promise of promises) {
+        await promise;
+    }
+}
+```
+
+We are still calling our varaible `promises` as this is what is really returned from `handle` and what is used by `async`/`await` syntax internally.
+
+You will probably agree that in this case the solution with `Promise` was much more readable. Like we described above we are first mapping set of commands to set of promises by calling `handle` method recursively and therefore they run in sequence. (We don't have a need for `map` callback function to be declared as `async` as it doesn't need to `await` for anything --- it receives `Promise` object from `handle` method and it is expected to return one.
+
+After mapping all commands to promises we need to wait for all of them to be resolved. Now it's time to use `await` in a `for ... of` loop --- there may be some commands that will take longer to be completed, but even if this loop will be temporarily blocked by them other commands will be already in progress.
+
+The order of awaiting doesn't matter here --- if some command runs quicker and completes before we use `await` on it there will be no need to pause the execution of our method.
+
+Finally, we don't need to return anything as `Promise` object is automatically created when we declare our method as `async`.
+
+Now we're done - we have compared implementations using `Promise` and `async`/`await`. In our case handling commands in parallel was more readable when we used `Promise`, while for running in sequence  `async`/`await` solution seems more suitable.
+
+This demonstrates that for writing succinct and readable code it's important to know both these techniques, to understand their innerworkings so that we can use either of them depending on a task at hand.
```

Comment:
```suggestion
This demonstrates that for writing succinct and readable code it's important to know both these techniques, to understand their inner workings so that we can use either of them depending on a task at hand.
```